File: top_interview_questions/1_easy/1_array/single_number/solution.py
# https://leetcode.com/problems/single-number/submissions/


def single_number(nums):
    xor_sum = 0

    for num in nums:
        xor_sum = xor_sum ^ num

    return xor_sum


if __name__ == '__main__':
    cases = [
        ([2, 2, 1], 1),
        ([4, 1, 2, 1, 2], 4)
    ]

    for case in cases:
        assert single_number(case[0]) == case[1]

File: top_interview_questions/1_easy/1_array/single_number/solution.go
package main

import "fmt"

func main() {
	fmt.Println(singleNumber([]int{1, 2, 2, 3}))    // expected 2
	fmt.Println(singleNumber([]int{2, 3, 4, 5, 5})) // expected 5
}

func singleNumber(nums []int) int {
	xorSum := 0

	for _, num := range nums {
		xorSum ^= num
	}

	return xorSum
}
